### PR TITLE
Add mage training arena plugin

### LINKED ISSUE #2218: Mage Training Arena plugin
In the mta alchemy room, the items are always in the same order in a clockwise direction

Boot
Shield 
Helm
Gem
Sword
Empty * 3

I was wondering if it was possible, so if you searched and found and item during that current rotation (40 secs or so), it could display the locations of the rest of the items on the alchemy cupboards, perhaps highlighting the one worth 30 coins.

### LINKED ISSUE #3257: Mage Training Arena Graveyard plugin, Bones Counter
Can someone do a Graveyard plugin for the Mage training arena which counts the equivalent bananas to bones, thank you.

### PR DESCRIPTION
**Pull request by Toocanzs**
Adds an overlay to show where items appear in the cabinets of the MTA alchemy room.

Also displays the total number of fruits generated from the bones in your inventory in the graveyard room.

https://i.gyazo.com/23dec938080015a8248a56f14edb9304.gif

Closes #2218 and closes #3257

### CODE REVIEW

**Comment by psikoi** [comment]
Is there other rooms added? can you include images?

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomCabinet.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package net.runelite.client.plugins.magetrainingarena;
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomCabinet.java]: Add copyright to every new file

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomItem.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package net.runelite.client.plugins.magetrainingarena;
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomItem.java]: Copyright

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomCabinet.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+public enum AlchemyRoomCabinet
+{
+	WEST_0(23684, 23685),
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomCabinet.java]: Are these object IDs not in `ObjectID.java`?

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomCabinet.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+	int closedObjectId;
+	int openObjectId;
+	AlchemyRoomCabinet(int closedObjectId, int openObjectId)
+	{
+		this.closedObjectId = closedObjectId;
+		this.openObjectId = openObjectId;
+	}
+
+	static AlchemyRoomCabinet getCabinetById(int objectId)
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomCabinet.java]: Static block should be before constructor

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	AlchemyRoomOverlay(Client client, MageTrainingArenaPlugin plugin)
+	{
+		this.client = client;
+		this.plugin = plugin;
+		setPosition(OverlayPosition.DYNAMIC);
+		setLayer(OverlayLayer.ABOVE_SCENE);
+		setPriority(OverlayPriority.LOW);
+	}
+
+
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java]: Too many newlines here

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import net.runelite.api.Client;
+import net.runelite.api.GameObject;
+import net.runelite.api.coords.LocalPoint;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayPriority;
+import net.runelite.client.ui.overlay.OverlayUtil;
+
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java]: Remove new line in imports

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomCabinet.java (hunk 1 above)
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomCabinet.java]: License header

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java (hunk 5 above)
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java]: Single blank line between blocks

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomItem.java (hunk 2 above)
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomItem.java]: License header

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package net.runelite.client.plugins.magetrainingarena;
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java]: License header

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/GraveyardBone.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package net.runelite.client.plugins.magetrainingarena;
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/GraveyardBone.java]: License header

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+				BufferedImage image = plugin.getCabinetItemImage(cabinet.getId());
+				LocalPoint localLoc = LocalPoint.fromWorld(client, cabinet.getWorldLocation());
+				if (image == null || localLoc == null)
+					continue;
+				OverlayUtil.renderImageLocation(client, graphics, localLoc, image, 100);
+			}
+		}
+		return null;
+	}
+
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java]: No blank lines at end of class

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/GraveyardBoneCounter.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package net.runelite.client.plugins.magetrainingarena;
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/GraveyardBoneCounter.java]: License header

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/GraveyardBoneCounter.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+import net.runelite.client.ui.overlay.infobox.Counter;
+
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/GraveyardBoneCounter.java]: No need for newline here

Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/GraveyardBoneCounter.java]: No blank lines in imports

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+	@Getter
+	private GameObject[] cabinets;
+
+	@Getter
+	private int totalFruitFromBones = 0;
+
+	private GraveyardBoneCounter counter;
+
+
+
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Too many newlines

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+		AlchemyRoomItem item = currentPermutation[permutationArrayIndex];
+
+		if (item == null)
+			return null;
+
+		BufferedImage image = itemManager.getImage(item.itemID);
+		return image;
+	}
+
+	private void AddBoneCounter()
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Empty method?

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+import net.runelite.api.queries.InventoryItemQuery;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.client.game.ItemManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.infobox.InfoBoxManager;
+import net.runelite.client.util.QueryRunner;
+
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Blank lines in imports

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+import java.util.Map;
+
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.ADAMANT_KITESHIELD;
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.ADAMANT_MED_HELM;
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.EMERALD;
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.LEATHER_BOOTS;
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.RUNE_LONGSWORD;
+
+@PluginDescriptor(
+		name = "Mage Training Arena"
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: There are one too many tabs in this line

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+	public void OnGameTick(GameTick event)
+	{
+		final GameObjectQuery query = new GameObjectQuery().idEquals(getCabinetObjectIds());
+		cabinets = queryRunner.runQuery(query);
+
+		Item[] items = queryRunner.runQuery(new InventoryItemQuery(InventoryID.INVENTORY));
+
+
+
+
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Just one newline is enough

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.ADAMANT_KITESHIELD;
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.ADAMANT_MED_HELM;
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.EMERALD;
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.LEATHER_BOOTS;
+import static net.runelite.client.plugins.magetrainingarena.AlchemyRoomItem.RUNE_LONGSWORD;
+
+@PluginDescriptor(
+		name = "Mage Training Arena"
+)
+
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Annotation goes directly above class

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+public class MageTrainingArenaPlugin extends Plugin
+{
+
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: No blank line at the beginning of class

**Comment by Toocanzs** [comment]
I've added an overlay for the alchemy room and an info box for the graveyard room. I don't plan on adding any other rooms as I don't think they would be very useful personally

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+		}
+	}
+
+	@Subscribe
+	public void OnMenuEntryOptionClicked(MenuOptionClicked event)
+	{
+		lastObjectIdClicked = event.getId();
+	}
+
+
```
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: No need for newlines here

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java (hunk 12 above)
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Only a single blank line

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java (hunk 18 above)
Comment (Kamielvf) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: No need for newline

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+	@Subscribe
+	public void OnChatMessage(ChatMessage event)
+	{
+		String message = event.getMessage();
+		if (message.contains("You found: "))
+		{
+			String item = message.split("You found: ")[1];
+			Integer arrayIndex = alchemyObjectIDToArrayIndexMap.get(lastObjectIdClicked);
+			if (arrayIndex == null)
+				return;
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Always use curly braces in if statements

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+
+			currentPermutation = getPermutation(item, arrayIndex);
+		}
+		if (message.equalsIgnoreCase("The cupboard is empty.") && currentPermutation != null)
+		{
+			Integer arrayIndex = alchemyObjectIDToArrayIndexMap.get(lastObjectIdClicked);
+			if (arrayIndex == null)
+				return;
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Always use curly braces in if statements

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+
+			//Reset if we start finding nothing and the current permutation doesn't match
+			if (currentPermutation[arrayIndex] != null)
+				currentPermutation = null;
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Always use curly braces in if statements

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+		}
+		//Reset if you re-enter the area
+		if (message.equalsIgnoreCase("You've entered the Alchemists' Playground."))
+			currentPermutation = null;
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Always use curly braces in if statements (various other in file)

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+			cabinetIds[i * 2] = cabinet.closedObjectId;
+			cabinetIds[i * 2 + 1] = cabinet.openObjectId;
+		}
+
+		return cabinetIds;
+	}
+
+	public BufferedImage getCabinetItemImage(int objectId)
+	{
+
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Remove blank line

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java (hunk 13 above)
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Useless method

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+	@Subscribe
+	public void OnGameTick(GameTick event)
+	{
+		final GameObjectQuery query = new GameObjectQuery().idEquals(getCabinetObjectIds());
+		cabinets = queryRunner.runQuery(query);
+
+		Item[] items = queryRunner.runQuery(new InventoryItemQuery(InventoryID.INVENTORY));
+
+
+
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Only one blank line

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+				infoBoxManager.removeInfoBox(counter);
+		}
+	}
+
+	@Subscribe
+	public void OnMenuEntryOptionClicked(MenuOptionClicked event)
+	{
+		lastObjectIdClicked = event.getId();
+	}
+
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: Remove blank lines at end of file

**Comment by deathbeam** [comment]
Telegrab room next tile to stand would be nice imo. Even though it is a bit obvious, but still can make it less annoying.

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+	public Dimension render(Graphics2D graphics)
+	{
+		GameObject[] cabinets = plugin.getCabinets();
+		if (cabinets != null)
+		{
+			for (GameObject cabinet : cabinets)
+			{
+				BufferedImage image = plugin.getCabinetItemImage(cabinet.getId());
+				LocalPoint localLoc = LocalPoint.fromWorld(client, cabinet.getWorldLocation());
+				if (image == null || localLoc == null)
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/AlchemyRoomOverlay.java]: Curly braces are required on all if statements.

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 28]
```diff
@@ -0,0 +1,263 @@
...
+		return cabinetIds;
+	}
+
+	public BufferedImage getCabinetItemImage(int objectId)
+	{
+		int permutationArrayIndex = alchemyObjectIDToArrayIndexMap.get(objectId);
+
+		if (currentPermutation == null)
+			return null;
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: braces

File: runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java [hunk 29]
```diff
@@ -0,0 +1,263 @@
...
+		AlchemyRoomItem item = currentPermutation[permutationArrayIndex];
+
+		if (item == null)
+			return null;
```
Comment (MESLewis) [inline runelite-client/src/main/java/net/runelite/client/plugins/magetrainingarena/MageTrainingArenaPlugin.java]: braces

**Comment by Toocanzs** [comment]
I've changed my mind and added a telekinetic room overlay. 

https://gfycat.com/HarmoniousKindFruitbat

**Comment by deathbeam** [comment]
Implemented in #3596

**Pull request closed without merging**